lib/executor/variable_looping_vus.go

```diff
@@ -439,10 +297,6 @@ func (vlvc VariableLoopingVUsConfig) GetExecutionRequirements(es *lib.ExecutionS
 	steps := vlvc.getRawExecutionSteps(es, false)
 
 	executorEndOffset := sumStagesDuration(vlvc.Stages) + time.Duration(vlvc.GracefulStop.Duration)
-	// Handle graceful ramp-downs, if we have them
-	if vlvc.GracefulRampDown.Duration > 0 {
-		steps = vlvc.reserveVUsForGracefulRampDowns(steps, executorEndOffset)
```

Comment:
Erm... This seems wrong... :confused: `GetExecutionRequirements()` **must** take into account `gracefulRampDown` values in the VU requirements it returns. Otherwise, some other executor will be expecting to reuse a VU that's still being used by the `variable-looping-vus` executor... This is what I wrote at the top of the docs as well:
```go
// If the graceful ramp-downs are enabled, we need to reserve any VUs that may
// potentially have to finish running iterations when we're scaling their number
// down. This would prevent attempts from other executors to use them while the
// iterations are finishing up during their allotted gracefulRampDown periods.
```

So, unless I'm missing something,`reserveVUsForGracefulRampDowns()` needs to stay... It's *primary* purpose is to be used here, so that k6 [can know how many VUs are needed to run the script at any given time](https://github.com/loadimpact/k6/blob/8d0271379dd3bcbcbd00dd660a9e15890ac2533f/lib/executors.go#L242-L264). Using it in the `Run()` method was a bonus, since we need to do basically the same calculation there for the hard stops...

Reply:
> Otherwise, some other executor will be expecting to reuse a VU that's still being used by the variable-looping-vus executor...

Ah, right, I knew this was too naive to work. :D

I didn't actually test this together with other executors to see this behavior.

I'll look into this a bit, but I really wanted to avoid the modification of the original "raw" execution plan. It would be simpler if there was one "true" execution plan, taking into account all configurable timeouts, and only that one is used for execution. I'm sure there's a reason for the distinction I'll soon discover that can't be removed.

Reply:
> I really wanted to avoid the modification of the original "raw" execution plan

Why? Don't think of it as a modification of the execution plan, but rather as a 2-step calculation to get the end result, with the side benefit that the result from the middle step is actually useful for other purposes...

> It would be simpler if there was one "true" execution plan, taking into account all configurable timeouts, and only that one is used for execution.

Would it really be simpler? I'd love that to be the case, but from my perspective, in order to properly and _efficiently_ manage the VUs, you need to know when to gracefully stop a VU, and when to completely stop it. And while I might be more than a bit biased, but I think that is very nicely achieved with the two channels, without the need for spawning new goroutines :wink:

Reply:
> Why? Don't think of it as a modification of the execution plan, but rather as a 2-step calculation to get the end result, with the side benefit that the result from the middle step is actually useful for other purposes...

Why is the 2-step calculation needed at all? Could we generate a single execution plan that takes all timeouts into consideration and returns both the amount of planned active VUs and the amount of planned "standby" VUs based on `gracefulRampDown` at specific time offsets? Do we need the "raw" execution plan at all? Even if we do, whatever needs it could just ignore the graceful events in it. This way we would stream only one execution plan and it would be much simpler overall.

> Would it really be simpler? I'd love that to be the case, but from my perspective, in order to properly and efficiently manage the VUs, you need to know when to gracefully stop a VU, and when to completely stop it.

Yes, I think it would be simpler. Consider the following:

```golang
[]github.com/loadimpact/k6/lib.ExecutionStep len: 32, cap: 32, [
        {TimeOffset: 0, ActiveVUs: 0, StandByVUs: 0, ...},
        {TimeOffset: 1000000000, ActiveVUs: 1, StandByVUs: 0, ...},
        {TimeOffset: 2000000000, ActiveVUs: 2, StandByVUs: 0, ...},
        {TimeOffset: 3000000000, ActiveVUs: 3, StandByVUs: 0, ...},
        {TimeOffset: 4000000000, ActiveVUs: 4, StandByVUs: 0, ...},
        {TimeOffset: 5000000000, ActiveVUs: 0, StandByVUs: 5, ...},
        {TimeOffset: 1000000000, ActiveVUs: 0, StandByVUs: 0, ...},
        ...
```

In this case:
- `ActiveVUs`: VUs actively running iterations.
- `StandByVUs`: VUs that are in the process of shutting down, awaiting iterations to complete, or in "stand-by" for lack of a better term. We could also use this in reverse to signal to the engine at which point to initialize more VUs and have them in "stand-by" for future use.
  
So this execution plan will ramp up to 5 VUs at t=5, immediately call `gracefulStop()` on them at the same tick, and finally call `hardStop()` on them at t=10.

If we generate this once then it's easier to reason about and the execution itself becomes a straightforward interpretation of events, instead of requiring [workarounds](https://github.com/loadimpact/k6/blob/1cd657f32740070155e482075f22e88c247b5010/lib/executor/variable_looping_vus.go#L593-L597) or potentially costly goroutines.

Reply:
> Why is the 2-step calculation needed at all? 

Maybe it's not needed, but I don't understand why you're so focused on it, it is just a useful implementation detail... The 2-pass algorithm for calculating the graceful rampdown times seemed intuitive to me, that's why I wrote it, and I made sure to document it fairly extensively. It's certainly simpler than the alternative of cramming everything together into a single huge and complicated method. And I don't understand why it matters so much whether you "zip" the two slices into a single one with `ActiveVUs` and `StandByVUs`, or you keep them separate...

Personally, I think that if the small fix for the `if` condition fixed the bug, you should just leave it at that, add a test, and write up your ideas for improving this in a separate `low prio` issue, for us to deal with in the future. Since none of these implementation details is exported outside of the type, we can always come back and change them later, but there are much more important missing and bugs things remaining to be done in #1007.

Reply:
> Maybe it's not needed, but I don't understand why you're so focused on it

Because I'd prefer to reduce complexity if there's a simpler alternative that wouldn't take much time to implement (unless I'm underestimating the amount of work, which is likely).

> The 2-pass algorithm for calculating the graceful rampdown times seemed intuitive to me, that's why I wrote it, and I made sure to document it fairly extensively.

Documentation is not "free". :) It still takes effort to understand and fit into one's own understanding of the program (and spot desync "bugs"). I appreciate the details of what you wrote, but if we could get rid of some of it I would see it as an improvement. That is, the less algorithm to describe, the better.


> Personally, I think that if the small fix for the `if` condition fixed the bug, you should just leave it at that, add a test and write up your ideas for improving this in a separate low prio issue, for us to deal with in the future.

Sure, this was unplanned. I think there's value in exploring it now, but OK, I'll close this PR, create an issue, and a PR for the fix.

Thanks for looking into it!

Reply:
There seems to be some misunderstanding or conflation of the algorithms here. There are 2 distinct algorithms:
1. The one that determines how many VUs would be needed at any given point in time, so we can satisfy the [`GetExecutionRequirements()` method in the interface](https://github.com/loadimpact/k6/blob/1cd657f32740070155e482075f22e88c247b5010/lib/executors.go#L97-L100). This **needs** to account for `gracefulRampDown` values. 
2. The one that actually starts and gracefully/fully stops VUs in `Run()`

It just so happens that the intermediary step of the first algorithm is useful to the second. To me, that somewhat simplifies everything. I don't particularly care if for algorithm 2. we express its steps in 2 separate slices or in a single slice with 2-part elements... The problem is that, so far, you haven't proposed a simpler alternative to 1, you just deleted an essential part of it :wink:   

> Documentation is not "free". :) 

I very much understand that, believe me. The problem is that the actual algorithm in 1) to calculate `gracefulRampDowns` isn't simple... It essentially has to look into the future, and that's why the current 2-pass approach was chosen - it seemed the simplest one conceptually, with a side-benefit of having a usable intermediate state...